scripts/gaps.py

```diff
@@ -0,0 +1,310 @@
+        remaining = {key: query[key] for key in query if key != 'time'}
+
+        for interval in divide(query['time'], time_divs):
+            yield dict(time=interval, **remaining)
+    else:
+        yield query
+
+
+def find_gaps(datacube, products, query, time_divs=None):
+    """ Summary of gaps in the `products` compared pairwise. """
+    products = list(set(products))
+    assert len(products) != 0 and len(products) != 1, "no products to compare"
+
+    grid_workflow = GridWorkflow(datacube.index, product=products[0])
+
+    def mismatch_summary(product1, product2):
+        """ Summarize mismatch info into a dictionary. """
+        subqueries = subdivide_time_domain(time_divs=time_divs,
+                                           **query)
+        left, right = distribute(datacube,
+                                 product1, product2, grid_workflow,
+                                 subqueries)
+        return {
+            'products': [left.product, right.product],
+            left.product: left.summary(),
+            right.product: right.summary()
+        }
+
+    return [mismatch_summary(product1, product2)
+            for product1, product2 in combinations(products, 2)]
+
+
+def str_to_date(spec, shift=0):
+    """ Parse :class:`str` specification to :class:`datetime`. """
+    parts = [int(x) for x in spec.split('-')]
+
+    try:
+        year, month, day = parts
+        return datetime(year, month, day) + relativedelta(days=shift)
+    except ValueError:
+        pass
+
+    try:
+        year, month = parts
+        return datetime(year, month, 1) + relativedelta(months=shift)
+    except ValueError:
+        year = parts[0]
+        return datetime(year, 1, 1) + relativedelta(years=shift)
+
+
+def time_query(start_date, end_date):
+    """ Time range of query. """
+    if start_date is None:
+        assert end_date is None, "no start date given"
+        return {}
+    else:
+        start = str_to_date(start_date)
+        if end_date is None:
+            end = datetime.now()
+        else:
+            # shift=1 picks up the last moment of that date spec
+            end = str_to_date(end_date, shift=1)
+        return {'time': Range(start, end)}
+
+
+# pylint: disable=no-value-for-parameter
+@click.command(help=__doc__)
+@click.argument('products', nargs=-1, type=str)
+@click.option('--output-file', '-o',
+              type=click.File('wt'),
+              default='-',
+              help="output .yaml file to write report to (default: stdout)")
+@click.option('--start-date', type=str,
+              help="start of date range (YYYY, YYYY-MM, or YYYY-MM-DD)")
+@click.option('--end-date', type=str,
+              help="end of date range (YYYY, YYYY-MM, or YYYY-MM-DD)")
+@click.option('--time-divs', type=int,
+              help="divide the date range (computational)")
+def main(products, output_file, start_date, end_date, time_divs):
+    """ Entry point. """
+    datacube = Datacube(app='find-those-gaps')
```

Comment:
Could we display usage information if no products are supplied?
:o)

Reply:
done?